src/Common/Auth/OpenIDConnect/Grant/CustomClientCredentialsGrant.php

```diff
@@ -0,0 +1,325 @@
+                }
+            } catch (CannotDecodeContent | InvalidTokenStructure | UnsupportedHeaderFound $exception) {
+                $this->logger->error(
+                    "CustomClientCredentialsGrant->getClientCredentials() failed to parse token",
+                    ['exceptionMessage' => $exception->getMessage()]
+                );
+                throw OAuthServerException::invalidClient($request);
+            }
+        } else {
+            throw OAuthServerException::invalidRequest("client_assertion_type", "assertion type is not supported");
+        }
+        return null;
+    }
+
+    /**
+     * Validate the client.
+     *
+     * @param ServerRequestInterface $request
+     *
+     * @throws OAuthServerException
+     *
+     * @return ClientEntityInterface
+     */
+    protected function validateClient(ServerRequestInterface $request)
+    {
+        // skip everything else for now.
+        list($clientId) = $this->getClientCredentials($request);
+
+
+        // grab the client
+        $client = $this->getClientEntityOrFail($clientId, $request);
+
+        // TODO: @adunsulag check with @bradymiller or @sjpadgett on whether this belongs in ClientRepository since we
+        // are type checking it against ClientEntity.  Currently all the JWK validation stuff is centralized in this
+        // grant... but knowledge of the client entity is inside the ClientRepository, either way I don't like the
+        // class cohesion problems this creates.
+        if (!($client instanceof ClientEntity)) {
+            $this->logger->error("CustomClientCredentialsGrant->validateClient() client returned was not a valid ClientEntity ", ['client' => $clientId]);
+            throw OAuthServerException::invalidClient($request);
+        }
+
+        // validate everything to do with the JWT...
+
+        // @see https://tools.ietf.org/html/rfc7523#section-3
+
+        // 1. iss claim required && must match iss URI of sender application
+        // 2.B sub claim required && must match client id
+        // 3. aud claim required && must match this oauth2 server.  Token endpoint may be used.  This value needs to
+        // be communicated out of band to registering applications (put in SMART App registration page
+        // 4. exp claim required && must be > (current time - skew time[60seconds]).
+        //      OPTIONAL choice reject extended period exp claim.  OpenEMR choice set to 24 hours
+        // 5. nbf claim if sent must be < (current time) or REJECT
+        // 6. iat claim if sent may be rejected.  OpenEMR choice set to 5 minutes
+        // 7. jti claim represents id of the web token, may be stored and checked against to prevent replay attacks
+        // 8. has other claims
+        // 9. MAC signature verification or REJECT
+        // MAC signature needs to use ES384 or RS384 signature verification @see https://tools.ietf.org/html/rfc7518
+        //
+        // 10. Reject any other invalid JWT per RFC 7519
```

Comment:
11. Reject if the wind from the north is greater than 5 mph or if keurig stops working :)

Reply:
Yep, about right.